iree/modules/hal/hal_module.cc

```diff
@@ -555,19 +558,18 @@ class HALModuleState final {
   StatusOr<vm::ref<iree_hal_descriptor_set_t>> DescriptorSetCreate(
       vm::ref<iree_hal_device_t> device,
       vm::ref<iree_hal_descriptor_set_layout_t> set_layout,
-      absl::Span<const std::tuple<int32_t, vm::ref<iree_hal_buffer_t>, int32_t,
-                                  int32_t>>
-          bindings) {
+      absl::Span<const int32_t> binding_ordinals,
+      absl::Span<const vm::ref<iree_hal_buffer_t>> binding_buffers,
+      absl::Span<const int32_t> binding_offsets,
+      absl::Span<const int32_t> binding_lengths) {
     IREE_TRACE_SCOPE0("HALModuleState::DescriptorSetCreate");
-    // TODO(benvanik): custom packing logic for the structs so that we marshal
-    // the VM args directly into them instead of needing this temporary array.
     absl::InlinedVector<iree_hal_descriptor_set_binding_t, 4> binding_structs(
-        bindings.size());
-    for (int i = 0; i < bindings.size(); ++i) {
+        binding_ordinals.size());
+    for (int i = 0; i < binding_ordinals.size(); ++i) {
       binding_structs[i] = {
-          std::get<0>(bindings[i]), std::get<1>(bindings[i]).get(),
-          static_cast<iree_device_size_t>(std::get<2>(bindings[i])),
-          static_cast<iree_device_size_t>(std::get<3>(bindings[i]))};
+          binding_ordinals[i], binding_buffers[i].get(),
```

Comment:
eol comments for each would help legibility (and also keep clang-format from globbing together)